cmd/hope/deploy.go

```diff
@@ -135,30 +138,60 @@ var deployCmd = &cobra.Command{
 					return err
 				}
 			case ResourceTypeDockerBuild:
-				// Strip the actual tag off the repo so that it defaults to the
-				//   latest.
-				tagSeparator := strings.LastIndex(resource.Build.Tag, ":")
-				pullImage := resource.Build.Tag
-				if tagSeparator != -1 {
-					pullImage = pullImage[:tagSeparator]
+				isCacheCommand := len(resource.Build.Source) != 0
+				isBuildCommand := len(resource.Build.Path) != 0
+
+				if isCacheCommand && isBuildCommand {
+					return errors.New(fmt.Sprintf("Docker build step %s cannot have a path and a source", resource.Name))
 				}
 
-				if err := docker.ExecDocker("pull", pullImage); err != nil {
-					// Maybe the image was pushed with the given tag.
-					// Maybe the tag is something like :stable.
-					// Hopefully we can grab a few layers at least.
-					if err := docker.ExecDocker("pull", resource.Build.Tag); err != nil {
-						log.Warn("Failed to pull existing images for ", pullImage, ". Maybe this image doesn't exist?")
-
-						// Don't return any errors here.
-						// If this is the first time this image is being
-						//   pushed, there will be nothing to pull, and
-						//   this will never succeed.
+				// TODO: Move these to constants somewhere
+				pullConstraintAlways := resource.Build.Pull == "always"
+				pullConstraintIfNotPresent := resource.Build.Pull == "if-not-present" || resource.Build.Pull == ""
+				
+				if !pullConstraintAlways && !pullConstraintIfNotPresent {
+					return errors.New(fmt.Sprintf("Unknown Docker image pull constraint: %s", resource.Build.Pull))
+				}
+
+				pullImage := ""
+
+				if isCacheCommand {
+					pullImage = resource.Build.Source
+				} else {
+					pullImage = resource.Build.Tag
+				}
+
+				ifNotPresentShouldPull := false
+				if pullConstraintIfNotPresent {
+					output, err := docker.GetDocker("images", pullImage)
+					if err != nil {
+						return err
+					}
+
+					if len(strings.Split(output, "\n")) == 1 {
```

Comment:
This is pretty flakey, but it Docker doesn't want to fail if the image isn't found when directly specified. Would probably be better to actually inspect contents?

Reply:
Wow, this lasted a whole 41 minutes. https://github.com/Eagerod/hope/commit/e13df8e9f33c2461a28ed87bb8f797c742d3ac8a